httpbakery/client.go

```diff
@@ -343,14 +344,47 @@ func (c *Client) setRequestBody(req *http.Request, body io.ReadSeeker) error {
 	if body == nil {
 		return nil
 	}
-	if req.Body == nil {
-		req.Body = ioutil.NopCloser(body)
-	} else {
-		_, err := body.Seek(0, 0)
-		if err != nil {
+	if req.Body != nil {
+		req.Body.Close()
+		if _, err := body.Seek(0, 0); err != nil {
 			return errgo.Notef(err, "cannot seek to start of request body")
 		}
 	}
+	// Always replace the body with a new readStopper so that
+	// the old request cannot interfere with the new request's reader.
+	req.Body = &readStopper{
```

Comment:
Concerned about this expectation set by the `net/http#NewRequest` godoc:

> If the provided body is also an io.Closer, the returned Request.Body is set to body and will be closed by the Client methods Do, Post, and PostForm, and Transport.RoundTrip.

If I use `httpbakery/Client.Do` with a body that's an `os.File`, how would that file now get closed? Would the caller now need to close the file? That might be surprising, if the caller is expecting net/http-like behavior.

I think using `Close()` to reset the reference to the underlying reader is a pretty clever workaround for the [net/http bug](https://github.com/golang/go/issues/12796)... but should we also keep track of the underlying reader across repeated calls to `DoWithBody` so that we can eventually close it for the caller?